### PR TITLE
Add and Event Matcher and Image and File events in the Matrix Connector

### PR DESCRIPTION
**Pull request by Cadair**
# Description

This is a follow on from #826 and I think the final major component of #719.

I am opening it here, because it might provide a little bit of context for #826 but is worth reviewing separately.

## Status
**READY FOR REVIEW**

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: opsdroid/core.py [hunk 1]
```diff
@@ -112,7 +114,13 @@ def critical(self, error, code):
     @staticmethod
     def handle_async_exception(loop, context):
         """Handle exceptions from async coroutines."""
-        _LOGGER.error(_("Caught exception"))
+        if "future" in context:
```
Comment (jacobtomlinson) [inline opsdroid/core.py]: This is interesting and seems unrelated to this PR. Happy for it to be here but you could comment a little on what it is and why it's here?

File: opsdroid/events.py [hunk 2]
```diff
@@ -1,46 +1,137 @@
...
+            raise NameError("An event subclass named {name} has already been "
+                            "defined. Event subclass names must be globally "
+                            "unique.".format(name=name))
+
+        mcls.event_registry[name] = cls
+
+        return cls
+
+
+class EventMetaClass(ABCMeta):
```
Comment (jacobtomlinson) [inline opsdroid/events.py]: Did you intend to declare this class twice?

File: opsdroid/events.py [hunk 3]
```diff
@@ -102,69 +192,103 @@ def __init__(self, user, room, connector,
...
-                opsdroid.stats["total_response_time"] + \
-                (now - self.created).total_seconds()
-            self.responded_to = True
+        return await super().respond(response)
 
-    async def react(self, emoji):
-        """React to this message with emoji using the specified connector.
 
-        Delays message if thinking delay present in config. file.
+class Typing(Event):  # pragma: nocover
```
Comment (jacobtomlinson) [inline opsdroid/events.py]: 🎉

File: opsdroid/extern/image_size/README.rst [hunk 4]
```diff
@@ -0,0 +1,121 @@
+
```
Comment (jacobtomlinson) [inline opsdroid/extern/image_size/README.rst]: Adding an external dependency as code in the core project is not ideal. I'm assuming the original author has not made this available on PyPI.

My preference would be to create a separate repo for this in the opsdroid org and push it to PyPI as a separate package, then include it in opsdroid with pip.

**Comment by jacobtomlinson** [comment]
I think I reviewed the wrong thing.

File: opsdroid/core.py (hunk 1 above)
Comment (Cadair) [reply opsdroid/core.py]: So this I put here because I wasn't getting the traceback for an error and this is the way I worked out how to make it log the full traceback.

**Comment by Cadair** [comment]
Yes, sorry this one builds on the other one, and I probably shouldn't have opened this one yet, I just have a PR opening reflex.

File: opsdroid/extern/image_size/README.rst (hunk 4 above)
Comment (Cadair) [reply opsdroid/extern/image_size/README.rst]: I opened this: https://github.com/scardine/image_size/issues/13 to ask for a version on PyPI, so I am hoping this is just a stop-gap

**Comment by Cadair** [comment]
This PR is now one commit in addition to #826 so you can review the last commit if you wish.

**Comment by jacobtomlinson** [comment]
You might want to rebase this onto the current master now that #826 has been merged.

**Comment by Cadair** [comment]
OK, so there are two library issues with this PR.

1) python-magic that I have used to determine the mimetype of files requires the c lib `libmagic`, which it does not package in its wheels.
2) I externed in a python only image size code because it's not on PyPI yet.

Thoughts, suggestions etc?

**Comment by jacobtomlinson** [comment]
I would be happy to fork that dependency into the opsdroid org and push it to pypi ourselves. Or just install using the git url.

**Comment by Cadair** [comment]
I am happy with that approach, or just whatever you think is best. I spent some time looking for the lightest way to do these things and this is what I came up with.

**Comment by jacobtomlinson** [comment]
This should get you moving again. https://pypi.org/project/opsdroid-get-image-size/

**Comment by Cadair** [comment]
So an update on this from memory:

1. We need to work out how to handle the `image_size` dependency I have externed into this PR.
2. We need to work through the issues of `python-magic` needing the `libmagic` C library on all platforms.
3. I need to write some tests.

**Comment by jacobtomlinson** [comment]
I've published a fork of the `get_image_size` dependency to PyPI as [opsdroid-get-image-size](https://pypi.org/project/opsdroid-get-image-size/).

Is there an alternative to `python-magic`?

Yes tests please!

**Comment by Cadair** [comment]
> Is there an alternative to python-magic?

Not that I have found. I think a lot of the connectors are going to benefit from being able to identify the type of the files they are getting passed as a bytes object.

**Comment by jacobtomlinson** [comment]
What about https://pypi.org/project/puremagic/?

**Comment by Cadair** [comment]
That looks good. I will try it out.

**Comment by SolarDrew** [comment]
> Yes, sorry this one builds on the other one, and I probably shouldn't have opened this one yet, I just have a PR opening reflex.

It's a real PRoblem...

**Comment by Cadair** [comment]
@jacobtomlinson when I try and pip install `opsdroid-git-image-size` I get the following error:

```
Collecting opsdroid-get-image-size
  Downloading https://files.pythonhosted.org/packages/72/9f/664df7464feb2b2ba9a4c31678c2d143b83c6e6d2b4c64b4704acae154f2/opsdroid_get_image_size-0.2.0.tar.gz
    ERROR: Complete output from command python setup.py egg_info:
    ERROR: Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-install-if2jvqck/opsdroid-get-image-size/setup.py", line 16, in <module>
        LONG_DESCRIPTION = read_long_description()
      File "/tmp/pip-install-if2jvqck/opsdroid-get-image-size/setup.py", line 12, in read_long_description
        with codecs.open('HISTORY.rst', 'r', 'utf8') as history:
      File "/usr/lib/python3.7/codecs.py", line 898, in open
        file = builtins.open(filename, mode, buffering)
    FileNotFoundError: [Errno 2] No such file or directory: 'HISTORY.rst'
    ----------------------------------------
ERROR: Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-install-if2jvqck/opsdroid-get-image-size/
```

**Comment by jacobtomlinson** [comment]
@Cadair thanks for the PR to fix it. Should now be available in 0.2.1.

**Comment by Cadair** [comment]
Awesome thanks for sorting that so quickly.

**Comment by Cadair** [comment]
@jacobtomlinson I found a new and exciting bug :laughing: https://github.com/opsdroid/image_size/pull/5

**Comment by jacobtomlinson** [comment]
@Cadair try 0.2.2 ;)

File: opsdroid/message.py [hunk 5]
```diff
@@ -17,6 +17,9 @@ class Message(NewMessage):
     Deprecated. Use ``opsdroid.events.Message`` instead.
     """
 
+    # Skip registering this class with Event
+    _no_register = True
+
```
Comment (SolarDrew) [inline opsdroid/message.py]: Why not register `Message` with `Event`?

Comment (Cadair) [reply opsdroid/message.py]: because this is the old fallback message not the actual `Message` which is in `events.py`.

File: opsdroid/connector/matrix/connector.py [hunk 6]
```diff
@@ -230,6 +228,46 @@ def _get_formatted_message_body(message, body=None, msgtype="m.text"):
...
+        return {
+            "w": width,
+            "h": height,
+            "mimetype": await image.get_mimetype(),
+            "size": len(await image.get_file_bytes())
+        }
+
+    @register_event(File)
+    @register_event(Image)
+    async def _send_image(self, file_event):
```
[the remark this answers is not in the document]
Comment (SolarDrew) [reply opsdroid/connector/matrix/connector.py]: > `@register_event` won't catch subclasses?

and more importantly do we want it to?

[the remark this answers is not in the document]
Comment (Cadair) [reply opsdroid/connector/matrix/connector.py]: No, `register_event` matches exact type only, mainly for this reason.

File: opsdroid/message.py (hunk 5 above)
Comment (SolarDrew) [reply opsdroid/message.py]: Oh this is the old one, my bad.

**Comment by Cadair** [comment]
I hope :crossed_fingers: this is good to go now.

File: docs/matchers/event.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+## Example
+
+```python
+from opsdroid.skill import Skill
+from opsdroid.matchers import match_event
+from opsdroid.events import Image, Message
+
+class ShoutyImagesSkill(Skill):
+    @match_event(Image)
+    async def loudimage(opsdroid, config, event):
```
Comment (jacobtomlinson) [inline docs/matchers/event.md]: When using class based skills we do not take `opsdroid` and `config` in the methods. They are passed to the `__init__` of the `Skill` superclass and set as properties.

File: requirements.txt [hunk 8]
```diff
@@ -18,3 +18,5 @@ pycron==1.0.0
 pyyaml==4.2b1
 setuptools==41.0.1
 websockets==7.0
+puremagic==1.5
+opsdroid-get-image-size
```
Comment (jacobtomlinson) [inline requirements.txt]: Can we pin the version of this?

**Comment by Cadair** [comment]
:tada: :tada: Thanks!!
